test/v1alpha1/service.go

```diff
@@ -344,3 +393,194 @@ func IsServiceRoutesNotReady(s *v1alpha1.Service) (bool, error) {
+	if err != nil {
+		t.Fatalf("Failed to get Gateway %s/%s: %v", Namespace, GatewayName, err)
+	}
+
+	// Update its Spec
+	newGateway := curGateway.DeepCopy()
+	newGateway.Spec.Servers = servers
+
+	// Update the Gateway
+	gw, err := clients.SharedClient.NetworkingV1alpha3().Gateways(Namespace).Update(newGateway)
+	if err != nil {
+		t.Fatalf("Failed to update Gateway %s/%s: %v", Namespace, GatewayName, err)
+	}
+
+	var selectors []string
+	for k, v := range gw.Spec.Selector {
+		selectors = append(selectors, k+"="+v)
+	}
+	selector := strings.Join(selectors, ",")
+
+	// Restart the Gateway pods: this is needed because Istio without SDS won't refresh the cert when the secret is updated
+	pods, err := clients.KubeClient.Kube.CoreV1().Pods("istio-system").List(metav1.ListOptions{LabelSelector: selector})
+	if err != nil {
+		t.Fatalf("Failed to list Gateway pods: %v", err)
+	}
+
+	// TODO(bancel): there is a race condition here if a pod listed in the call above is deleted before calling watch below
+
+	var wg sync.WaitGroup
+	wg.Add(len(pods.Items))
+	wtch, err := clients.KubeClient.Kube.CoreV1().Pods("istio-system").Watch(metav1.ListOptions{LabelSelector: selector})
+	if err != nil {
+		t.Fatalf("Failed to watch Gateway pods: %v", err)
+	}
+	defer wtch.Stop()
+
+	done := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case event := <-wtch.ResultChan():
+				if event.Type == watch.Deleted {
+					wg.Done()
+				}
+			case <-done:
+				return
+			}
+		}
+	}()
+
+	err = clients.KubeClient.Kube.CoreV1().Pods("istio-system").DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: selector})
+	if err != nil {
+		t.Fatalf("Failed to delete Gateway pods: %v", err)
+	}
+
+	wg.Wait()
+	done <- struct{}{}
+}
+
+// setupHTTPS creates a self-signed certificate, installs it as a Secret and returns an *http.Transport
+// trusting the certificate as a root CA.
+func setupHTTPS(t *testing.T, kubeClient *ptest.KubeClient, hosts []string) {
+	t.Helper()
+
+	cert, key, err := generateCertificate(hosts)
+	if err != nil {
+		t.Fatalf("Failed to generate the certificate: %v", err)
+	}
+
+	rootCAs, _ := x509.SystemCertPool()
+	if rootCAs == nil {
+		rootCAs = x509.NewCertPool()
+	}
+
+	if ok := rootCAs.AppendCertsFromPEM(cert); !ok {
+		t.Fatalf("Failed to add the certificate to the root CA")
+	}
+
+	kubeClient.Kube.CoreV1().Secrets("istio-system").Delete("istio-ingressgateway-certs", &metav1.DeleteOptions{})
+	_, err = kubeClient.Kube.CoreV1().Secrets("istio-system").Create(&corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "istio-system",
+			Name:      "istio-ingressgateway-certs",
+		},
+		Type: corev1.SecretTypeTLS,
+		Data: map[string][]byte{
+			"tls.key": key,
+			"tls.crt": cert,
+		},
+	})
+	if err != nil {
+		t.Fatalf("Failed to set Secret %s/%s: %v", "istio-system", "istio-ingressgateway-certs", err)
+	}
+
+	HttpsTransportOption = func(transport *http.Transport) *http.Transport {
```

Comment:
I think this could be problematic when running tests in parallel because HttpsTransportOption is shared across each test case.
When `TestA` calls `setupHTTPS`, `HttpsTransportOption` is set to a value based on the cert generated for `TestA`. During running `TestA`, `HttpsTransportOption` could be set to another value by `TestB`, which could break `TestA`.

Reply:
Good point! I've updated the PR to use the map for `HttpsTransportOption` where the key is the test service names which will be unique. Please let me know if you think think will solve the problem.